doc/design.md

```diff
@@ -0,0 +1,136 @@
+available over HTTPS. It's capable of receiving metrics via statsd, and
+offers a simple Prometheus scraper as well. This document discusses the
+way in which all of these things interact.
+
+## Metrics
+
+Those familiar with standard operational metric terminology won't be
+surprised to learn that agentmon has a concept of counters and gauges.
+These behave very similarly to counters and gauges from
+the [statsd][statsd] "protocol". 
+
+That is to say, for a time period defined by the flush interval, the
+value of a named counter metric is the sum of all values reported for
+that name.
+
+For a gauge, in that same time period, the last value received for a
+given name is the value that will be reported.
+
+Outside of the statsd "protocol," agentmon is also capable of handling
+monotonically increasing counters as well. Internally, these are called
+"derived counters," and they are simply flushed as regular counters. 
+Each time a measurement is observed, the value added to the counter is
+equal to `{observed value} - {previously observed value}`.
+
+In the [Etsy statsd][etsy-statsd] implementation, counters, and timers
+can have an attached sample rate that is typically used to reduce
+observations sent to statsd, accepting, possibly, a bit less accuracy
+in reported observations. Statsd timers, however, are not handled in
+agentmon at the time of this writing. 
+
+### Why no timers?
+
+As it turns out, timers just haven't been necessary for the types of
+values we're collecting to date. However, there are also some
+challenges, and incompatibilities with the rest of the Heroku metrics
+infrastructure.
+
+First, the incompatibilities. The Heroku Metrics infrastructure makes
+use of a library called HDR Histogram, which can record the
+distribution of a large number of values at a relatively low cost. The
+cost is typically less than 1KiB in memory, but the serialization format
+is not standardized making it difficult for other agents needing to 
+report to the Heroku Application Metrics endpoints difficult.
+
+An alternative strategy would involve reporting multiple, related
+metrics, namely, sum of squares, sum, min, max, and count for each
+timer metric.  This would allow all of the reporting dynos' values to
+be fairly represented, and we could even derive standard deviation,
+and your typical average latency. However, average latency, and even
+standard deviation of latency isn't all that helpful. What's helpful
+are the tail latencies (p95, p99, p99.9), and often times the median
+latency. These _aren't_ derivable without knowing what the distribution
+looks like, which as we've previously stated, is not representable in
+an easily standardizable way.
+
+## Reporting Metrics to Heroku
+
+When started, the agentmon program expects a URL passed as an argument
+which speaks the protocol outlined below.
+
+While it'd certainly be possible to report metrics to other services,
+and via other means, the only reporter shipping with agentmon is a
```

Comment:
`While it'd certainly be possible to report metrics to other services via other means, the only reporter shipping with agentmon .... ` ditch the `and`, sentence fragment